File: tests/io/foreign_types.js
// Tree
// ====

// A balanced tree of the depth asked, its leaves 1.. left to right, every
// odd one Empty, as the C lays it.
function tree_make(depth) {
  let next = 0;
  const at = (d) => {
    if (d === 0) {
      next += 1;
      return next % 2 === 1 ? { $: "Empty" } : { $: "Leaf", v: next };
    }
    const l = at(d - 1);
    return { $: "Node", l, r: at(d - 1) };
  };
  return at(depth);
}

// Chain
// =====

function chain_make(n) {
  let c = { $: "End" };
  for (let i = 0; i < n; i += 1) {
    c = { $: "Cell", v: i, next: c };
  }
  return c;
}

File: tests/io/foreign_types.c
// Tree
// ====

// A balanced tree of the depth asked, its leaves 1.. left to right, every
// odd one Empty, laid as the program lays one: an Empty or a Leaf packed
// in its word, a Node a plain node (not sealed: the program's consuming
// match takes it as its own).
static Term tree_make_at(Env e, u32 depth, u32* next) {
  if (depth == 0) {
    *next += 1;
    return *next % 2 == 1 ? term_pak(CID_EMPTY, 0) : term_pak(CID_LEAF, *next);
  }
  Term l = tree_make_at(e, depth - 1, next);
  Term r = tree_make_at(e, depth - 1, next);
  return io_node(e, CID_NODE, l, r, 0);
}

Term tree_make_run(Env e, Term* f, IoWork* w) {
  u32 next = 0;
  return tree_make_at(e, (u32)f[0], &next);
}

// Chain
// =====

Term chain_make_run(Env e, Term* f, IoWork* w) {
  Term c = term_pak(CID_END, 0);
  for (u32 i = 0; i < (u32)f[0]; i += 1) {
    c = io_node(e, CID_CELL, i, c, 0);
  }
  return c;
}

static void __attribute__((constructor)) foreign_types_use(void) {
  io_eff(CID_TREE_MAKE, tree_make_run, 0);
  io_eff(CID_CHAIN_MAKE, chain_make_run, 0);
}
